File: lib/JFFNetwork/Categories/NSMutableURLRequest+CreateRequestWithURLParams.m
#import "NSMutableURLRequest+CreateRequestWithURLParams.h"

#import "JFFURLConnectionParams.h"

@implementation NSMutableURLRequest (CreateRequestWithURLParams)

+ (instancetype)mutableURLRequestWithParams:(JFFURLConnectionParams *)params
{
    NSInputStream *inputStream = params.httpBodyStreamBuilder?params.httpBodyStreamBuilder():nil;
    
    NSParameterAssert((!params.httpBody && !inputStream) || !(params.httpBody && inputStream));
    
    static const NSTimeInterval timeout = 60.;
    NSMutableURLRequest *request = [self requestWithURL:params.url
                                            cachePolicy:NSURLRequestReloadIgnoringLocalCacheData
                                        timeoutInterval:timeout];
    
    NSString *httpMethod = params.httpMethod ?: @"GET";
    if (!params.httpMethod && (params.httpBody || inputStream)) {
        httpMethod = @"POST";
    }
    
    [request setHTTPBodyStream:inputStream];
    if (params.httpBody)
        [request setHTTPBody:params.httpBody];
    
    [request setAllHTTPHeaderFields:params.headers];
    [request setHTTPMethod         :httpMethod    ];
    
    return request;
}

@end

File: iAsync.Swift/lib/JNetwork/JNetwork/Extensions/NSMutableURLRequest+CreateRequestWithURLParams.swift
//
//  NSMutableURLRequest+CreateRequestWithURLParams.swift
//  JNetwork
//
//  Created by Vladimir Gorbenko on 24.09.14.
//  Copyright (c) 2014 EmbeddedSources. All rights reserved.
//

import Foundation

extension NSMutableURLRequest {

     convenience init(params: JURLConnectionParams) {
    
        let inputStream = { () -> NSInputStream? in
            
            if let factory = params.httpBodyStreamBuilder {
                return factory()
            }
            return nil
        }()
    
        assert(!((params.httpBody != nil) && (inputStream != nil)))
    
        self.init(
            URL: params.url,
            cachePolicy: NSURLRequestCachePolicy.ReloadIgnoringLocalCacheData,
            timeoutInterval: 60.0)
        
        let httpMethod = { () -> String in
            
            if params.httpMethod == nil && (params.httpBody != nil || inputStream != nil) {
                
                return "POST"
            }
            return params.httpMethod ?? "GET"
        }()
        
        self.HTTPBodyStream = inputStream
        if let httpBody = params.httpBody {
            self.HTTPBody = params.httpBody
        }
        
        self.allHTTPHeaderFields = params.headers
        self.HTTPMethod          = httpMethod
    }
}
